### PR TITLE
[AIEX] Stuckmann.inter.mbb.ptrchains

### PR DESCRIPTION
**Pull request by F-Stuckmann**
Teach AIEClusterBaseAddress to reuse the tail of a dominating MBB's ptr-add chain when starting a new chain on the same base, instead of rematerialising the base every block.

The algorithm consists of 2 parts
1) Collect Ptr offsets at the end of MBBs.
2) Find dominating Ptr tail to start the base off that tail instead of a clean start.

### CODE REVIEW

**Comment by niwinanto** [comment]
@F-Stuckmann can we include tests from https://github.com/Xilinx/llvm-aie/pull/1008/changes#diff-f14e687e8be5d8f0f2625215a935d1099c7892d12dd0feeb3704036afe2c6598 as well.

**Comment by F-Stuckmann** [comment]
These unit-tests do not trigger AIEClusterBaseAddressing. 
For them to be found, we have to create 2 ptr_adds in each MBB, so i did that here. 

I also merged them into GlobalISel and renamed them for consistency with *cluster-base-address-*

`llvm/test/CodeGen/AIE/aie2ps/GlobalIsel/global-combiners/lag-reg-add-coalesce.mir` -> `llvm/test/CodeGen/AIE/GlobalISel/cluster-base-address-postinc-pre-ptr-coalesce.mir`

`llvm/test/CodeGen/AIE/aie2ps/GlobalIsel/global-combiners/lag-ptr-add.mir` has been converted and reduced to the edge cases in `llvm/test/CodeGen/AIE/GlobalISel/cluster-base-address-cross-block-reroot.mir` .
Since we now dont check Immediate size offsets, these tests have been reduced.

File: llvm/lib/Target/AIE/AIEClusterBaseAddress.cpp [hunk 1]
```diff
@@ -103,6 +110,50 @@ std::optional<int64_t> getConstOffset(const MachineInstr &MI,
...
+           "duplicate base record for this MBB");
+    Records[MBB][BaseReg] = {EndReg, Offset};
+  }
+
+  /// Walk MBB's idom chain upward and return the nearest dominating block's
+  /// record for BaseReg, if any.
+  std::optional<BlockEndOffset>
+  findDominatingBlockEndOffset(Register BaseReg, MachineBasicBlock &MBB) const {
+    auto *N = MDT.getNode(&MBB);
+    for (N = N ? N->getIDom() : nullptr; N; N = N->getIDom()) {
```
Comment (martien-de-jong) [inline llvm/lib/Target/AIE/AIEClusterBaseAddress.cpp]: nit: stepping to the root of the dominator tree can stretch the liveness dramatically.
We may be straddling function calls, suffering callee change effects

File: llvm/lib/Target/AIE/AIEClusterBaseAddress.cpp [hunk 2]
```diff
@@ -103,6 +110,50 @@ std::optional<int64_t> getConstOffset(const MachineInstr &MI,
...
+      auto BlockIt = Records.find(N->getBlock());
+
+      // This dominator published no chain tails at all; a higher dominator
+      // may still hold one for BaseReg, so keep climbing rather than exit.
+      if (BlockIt == Records.end())
+        continue;
+
```
Comment (martien-de-jong) [inline llvm/lib/Target/AIE/AIEClusterBaseAddress.cpp]: nit: we can break if the block of the basereg doesn't dominate anymore

File: llvm/lib/Target/AIE/AIEClusterBaseAddress.cpp [hunk 3]
```diff
@@ -103,6 +110,50 @@ std::optional<int64_t> getConstOffset(const MachineInstr &MI,
...
+    Records[MBB][BaseReg] = {EndReg, Offset};
+  }
+
+  /// Walk MBB's idom chain upward and return the nearest dominating block's
+  /// record for BaseReg, if any.
+  std::optional<BlockEndOffset>
+  findDominatingBlockEndOffset(Register BaseReg, MachineBasicBlock &MBB) const {
+    auto *N = MDT.getNode(&MBB);
+    for (N = N ? N->getIDom() : nullptr; N; N = N->getIDom()) {
+      auto BlockIt = Records.find(N->getBlock());
```
Comment (martien-de-jong) [inline llvm/lib/Target/AIE/AIEClusterBaseAddress.cpp]: Perhaps have a separate type for the inner map.

File: llvm/lib/Target/AIE/AIEClusterBaseAddress.cpp
Comment (martien-de-jong) [inline llvm/lib/Target/AIE/AIEClusterBaseAddress.cpp]: structural binding to key and value?

File: llvm/lib/Target/AIE/AIEClusterBaseAddress.cpp [hunk 4]
```diff
@@ -290,12 +368,38 @@ bool AIEClusterBaseAddress::processBasicBlock(MachineBasicBlock &MBB,
     if (shouldSkipChaining(RegAndUse.first, Instrs, MBB))
       continue;
 
-    // Build chain, breaking it (or restarting it) when necessary
+    // Snapshot the tail's absolute offset from BaseReg before buildChain
+    // rewrites operand(2) into a per-hop delta.
+    const std::optional<int64_t> TailAbsOffset =
+        getConstOffset(*Instrs.back(), *MRI);
```
Comment (martien-de-jong) [inline llvm/lib/Target/AIE/AIEClusterBaseAddress.cpp:401]: CHECK: we need to do this first, since chaining will spread the resulting offset across the chain.

File: llvm/lib/Target/AIE/AIEClusterBaseAddress.cpp [hunk 5]
```diff
@@ -290,12 +368,38 @@ bool AIEClusterBaseAddress::processBasicBlock(MachineBasicBlock &MBB,
     if (shouldSkipChaining(RegAndUse.first, Instrs, MBB))
       continue;
 
-    // Build chain, breaking it (or restarting it) when necessary
+    // Snapshot the tail's absolute offset from BaseReg before buildChain
+    // rewrites operand(2) into a per-hop delta.
+    const std::optional<int64_t> TailAbsOffset =
```
Comment (martien-de-jong) [inline llvm/lib/Target/AIE/AIEClusterBaseAddress.cpp]: nit: 'absolute' reminds me of 'absolute value'. I think it's just the offset of the tail from the base.

File: llvm/lib/Target/AIE/AIEClusterBaseAddress.cpp [hunk 6]
```diff
@@ -483,6 +587,107 @@ bool AIEClusterBaseAddress::buildChain(SmallVector<MachineInstr *, 8> &Instrs,
...
+}
+
+// True iff EndReg already feeds memory in MBB. A new chain rooted on it would
+// walk EndReg in place between the existing memory user and the new heads,
+// clobbering it.
+static bool anchorAlreadyHasMemoryUserInBlock(Register EndReg,
+                                              MachineBasicBlock &MBB,
+                                              const MachineRegisterInfo &MRI) {
+  for (const MachineInstr &UseMI : MRI.use_nodbg_instructions(EndReg))
+    if (UseMI.getParent() == &MBB && UseMI.mayLoadOrStore())
```
Comment (martien-de-jong) [inline llvm/lib/Target/AIE/AIEClusterBaseAddress.cpp]: nit: Store has a pointer use and a value use. If you don't distinguish them, I'm wondering why we restrict ourselves to checking for loads and stores only. Why not forbid any use.

**Review by niwinanto — COMMENTED** [review]
PR looks good to me, can we have the qor report as well?

File: llvm/lib/Target/AIE/AIEClusterBaseAddress.cpp (hunk 4 above)
Comment (F-Stuckmann) [reply llvm/lib/Target/AIE/AIEClusterBaseAddress.cpp:401]: yup

**Comment by F-Stuckmann** [comment]
Core_Compute_Insn_Count_perf_sorted_reduced
Core_PMSize_perf_sorted_reduced_absolute
Core_StackSize_perf_sorted_reduced_absolute

File: llvm/lib/Target/AIE/AIEClusterBaseAddress.cpp [hunk 7]
```diff
@@ -95,6 +95,12 @@ static cl::opt<bool> EnableCrossBlockChainReuse(
     cl::desc("Reuse the tail of a ptr-add chain produced in a strictly "
              "dominating MBB when starting a new chain on the same base."));
 
+static cl::opt<unsigned> CrossBlockReuseMaxDomSteps(
+    "aie-chain-addr-cross-block-reuse-max-steps", cl::Hidden, cl::init(8),
```
Comment (martien-de-jong) [inline llvm/lib/Target/AIE/AIEClusterBaseAddress.cpp:99]: I don't have hard numbers, but I think most of the dominator trees are less deep than 8.

File: llvm/lib/Target/AIE/AIEClusterBaseAddress.cpp [hunk 8]
```diff
@@ -136,22 +142,42 @@ class PtrChainEndTracker {
...
+      // No dominator above BaseReg's defining block can hold a record for it.
+      if (DomBlock == DefMBB)
+        break;
+
+      // Reusing an anchor from a higher dominator would stretch its live range
+      // across this block's call, exposing it to callee clobbers and spills.
+      const bool DomBlockHasCall = llvm::any_of(
+          *DomBlock, [](const MachineInstr &MI) { return MI.isCall(); });
+      if (DomBlockHasCall)
+        break;
```
Comment (martien-de-jong) [inline llvm/lib/Target/AIE/AIEClusterBaseAddress.cpp:180]: Note that there may be blocks in the live range that are not dominating us. We will not see the two branches of an intervening if-the-else

**Review by martien-de-jong — APPROVED** [review]
